src/decorators/grpc.decorator.ts

```diff
@@ -3,35 +3,184 @@
+ * import {
+ *   TestRequest,
+ *   TestResponse,
+ *   Greeter,
+ *   protoMetadata,
+ * } from '../protos-ts/greeter';
  *
- * class GreeterCtrl implements Greeter.Service {
- *   @grpc(Greeter.SayHello)
- *   public sayHello(request: HelloRequest): HelloResponse {
- *     return { message: 'Hello ' + call.request.name };
+ * class GreeterCtrl implements Greeter {
+ *   // Tell LoopBack that this is a Service RPC implementation
+ *   @grpc(protoMetadata.fileDescriptor)
+ *   async unaryTest(request: TestRequest): Promise<TestResponse> {
+ *     return {
+ *       message: 'Hello ' + request.name,
+ *     };
  *   }
  * }
  * ```
  */
-export function grpc(spec: GrpcMethod) {
-  return MethodDecoratorFactory.createDecorator(GRPC_METHODS, spec);
+export function grpc(
+  proto: FileDescriptorProto,
+  options?: {
+    /**
+     * Defaults to /(Ctrl|Controller)$/
+     */
+    controllerNameRegex?: RegExp;
+  },
+) {
+  return function <
+    GrpcRequest,
+    GrpcResponse,
+    M1 extends GrpcUnaryMethod<GrpcRequest, GrpcResponse>,
+    M2 extends GrpcServerStreamingMethod<GrpcRequest, GrpcResponse>,
+    M3 extends GrpcClientStreamingMethod<GrpcRequest, GrpcResponse>,
+    M4 extends GrpcBidiStreamingMethod<GrpcRequest, GrpcResponse>,
+    T extends
+      | TypedPropertyDescriptor<M1>
+      | TypedPropertyDescriptor<M2>
+      | TypedPropertyDescriptor<M3>
+      | TypedPropertyDescriptor<M4>,
+  >(
+    target: Object,
+    propertyKey: string | symbol,
+    descriptor: T extends TypedPropertyDescriptor<infer F>
+      ? F extends M1
+        ? TypedPropertyDescriptor<GrpcUnaryMethod<GrpcRequest, GrpcResponse>>
+        : F extends M2
+        ? TypedPropertyDescriptor<
+            GrpcServerStreamingMethod<GrpcRequest, GrpcResponse>
+          >
+        : F extends M3
+        ? TypedPropertyDescriptor<
+            GrpcClientStreamingMethod<GrpcRequest, GrpcResponse>
+          >
+        : TypedPropertyDescriptor<F>
+      : T,
+  ):
+    | (T extends TypedPropertyDescriptor<infer F>
+        ? F extends M1
+          ? TypedPropertyDescriptor<GrpcUnaryMethod<GrpcRequest, GrpcResponse>>
+          : F extends M2
+          ? TypedPropertyDescriptor<
+              GrpcServerStreamingMethod<GrpcRequest, GrpcResponse>
+            >
+          : F extends M3
+          ? TypedPropertyDescriptor<
+              GrpcClientStreamingMethod<GrpcRequest, GrpcResponse>
+            >
+          : TypedPropertyDescriptor<F>
+        : T)
+    | void {
+    if (!descriptor.value) return;
+
+    const methodName = descriptor.value.name;
+    const spec = getLoopbackMetadataFromMethodDescriptor(
+      findMethodInFileDescriptor(
+        proto,
+        target.constructor.name.replace(
+          options?.controllerNameRegex ?? new RegExp(/(Ctrl|Controller)$/, 'g'),
+          '',
+        ),
+        `${methodName[0].toUpperCase()}${
+          methodName.length ? methodName.slice(1) : ''
+        }`,
+      ),
+    );
```

Comment:
Would it be possible to enable explicit declaration of the gRPC method name via the decorator instead (i.e. Like how it was before)?

I understand the rationale for inferring this automatically, but IMO doing things auto-magically has its pitfalls of being  more difficult to debug. It's also not as aligned with the "LoopBack 4 approach" of giving developers control and flexibility.

Reply:
Well yes. I had done this because of limitations of the compiler.
I am currently trying to do something else, more explicit.

Reply:
I have modified the decorator. It now takes an object as argument, with all the necessary properties. I have also provided a helper function which takes the fileDescriptor and the method definition.